test/Microsoft.ML.Tests/Scenarios/Api/EstimatorPipe.cs

```diff
@@ -0,0 +1,337 @@
+                verWrittenCur: 0x00010001, // Initial
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: TransformerChain.LoaderSignature);
+        }
+
+    }
+
+    public static class TransformerChain
+    {
+        public const string LoaderSignature = "TransformerChain";
+
+        public static ITransformer Create(IHostEnvironment env, ModelLoadContext ctx) => new TransformerChain<ITransformer>(env, ctx);
+    }
+
+    public sealed class CompositeReader<TSource, TLastTransformer> : IDataReader<TSource>
+        where TLastTransformer : class, ITransformer
+    {
+        public readonly IDataReader<TSource> Reader;
+        public readonly TransformerChain<TLastTransformer> Transformer;
+
+        public CompositeReader(IDataReader<TSource> reader, TransformerChain<TLastTransformer> transformerChain = null)
+        {
+            Contracts.CheckValue(reader, nameof(reader));
+            Contracts.CheckValueOrNull(transformerChain);
+            Reader = reader;
+            Transformer = transformerChain ?? new TransformerChain<TLastTransformer>();
+        }
+
+        public IDataView Read(TSource input)
+        {
+            var idv = Reader.Read(input);
+            idv = Transformer.Transform(idv);
+            return idv;
+        }
+
+        public ISchema GetOutputSchema()
+        {
+            var s = Reader.GetOutputSchema();
+            s = Transformer.GetOutputSchema(s);
+            return s;
+        }
+
+        public void SavePipeline(IHostEnvironment env, Stream outputStream)
+        {
+            using (var ch = env.Start("Saving model"))
+            {
+                using (var rep = RepositoryWriter.CreateNew(outputStream, ch))
+                {
+                    ch.Trace("Saving data reader");
+                    ModelSaveContext.SaveModel(rep, Reader, "Reader");
+
+                    ch.Trace("Saving transformer chain");
+                    ModelSaveContext.SaveModel(rep, Transformer, "TransformerChain");
+                    rep.Commit();
+                }
+            }
+        }
+    }
+
+    public static class CompositeReader
+    {
+        public static CompositeReader<IMultiStreamSource, ITransformer> LoadPipeline(IHostEnvironment env, Stream stream)
+        {
+            using (var rep = RepositoryReader.Open(stream, env))
+            {
+                ModelLoadContext.LoadModel<IDataReader<IMultiStreamSource>, SignatureLoadModel>(env, out var reader, rep, "Reader");
+                ModelLoadContext.LoadModel<TransformerChain<ITransformer>, SignatureLoadModel>(env, out var transformerChain, rep, "TransformerChain");
+                return new CompositeReader<IMultiStreamSource, ITransformer>(reader, transformerChain);
+            }
+        }
+    }
+
+    [Flags]
+    public enum TransformerScope
+    {
+        None = 0,
+        Training = 1 << 0,
+        Testing = 1 << 1,
+        Scoring = 1 << 2,
```

Comment:
>Scoring [](start = 8, length = 7)

considering what usually testing is scoring + get metrics, can we put scoring above of testing?

Reply:
I was looking more chronologically at it: first, you need to train the model. Then, you are going to test it, and then, finally, when you are satisfied with what you've got, you package it and use it for prediction / scoring.

---
In reply to: [210998340](https://github.com/dotnet/machinelearning/pull/688#discussion_r210998340) [](ancestors = 210998340)